lib/ansible/modules/windows/win_domain.ps1

```diff
@@ -56,99 +86,188 @@ $installed = Ensure-Prereqs
-        if ($_.Exception.ExitCode -eq 15) {
+        if ($domain_netbios_name) {
+            $install_params.DomainNetBiosName = $domain_netbios_name
+        }
+
+        if ($null -ne $create_dns_delegation) {
+            $install_params.CreateDnsDelegation = $create_dns_delegation
+        }
+
+        if ($domain_mode) {
+            $install_params.DomainMode = $domain_mode
+        }
+
+        if ($forest_mode) {
+            $install_params.ForestMode = $forest_mode
+        }
+
+        $iaf = $null
+        try {
+            $iaf = Install-ADDSForest @install_params
+        } catch [Microsoft.DirectoryServices.Deployment.DCPromoExecutionException] {
+            # ExitCode 15 == 'Role change is in progress or this computer needs to be restarted.'
+            # DCPromo exit codes details can be found at https://docs.microsoft.com/en-us/windows-server/identity/ad-ds/deploy/troubleshooting-domain-controller-deployment
+            if ($_.Exception.ExitCode -eq 15) {
+                $result.reboot_required = $true
+            } else {
+                $module.FailJson("Failed to install ADDSForest with DCPromo: $($_.Exception.Message)")
+            }
+        }
+
+        if ($check_mode) {
+            # the return value after -WhatIf does not have RebootRequired populated
+            # manually set to True as the domain would have been installed
             $result.reboot_required = $true
-        } else {
-            Fail-Json -obj $result -message "Failed to install ADDSForest with DCPromo: $($_.Exception.Message)"
+        } elseif ($null -ne $iaf) {
+            $result.reboot_required = $iaf.RebootRequired
+
+            # The Netlogon service is set to auto start but is not started. This is
+            # required for Ansible to connect back to the host and reboot in a
+            # later task. Even if this fails Ansible can still connect but only
+            # with ansible_winrm_transport=basic so we just display a warning if
+            # this fails.
+            try {
+                Start-Service -Name Netlogon
+            } catch {
+                $module.Warn("Failed to start the Netlogon service after promoting the host, Ansible may be unable to connect until the host is manually rebooting: $($_.Exception.Message)")
+            }
         }
     }
+} else {
+    # parent domain is defined, creating a new child domain
+    $domain_fqdn = "$dns_domain_name.$parent_domain_name"
+
+    $domain = $null
+    try {
+        $domain_context =  New-Object -TypeName System.DirectoryServices.ActiveDirectory.DirectoryContext -ArgumentList Domain, $domain_fqdn
+        $domain = [System.DirectoryServices.ActiveDirectory.Domain]::GetDomain($domain_context)
+    } catch [System.DirectoryServices.ActiveDirectory.ActiveDirectoryObjectNotFoundException] {
+    } catch [System.DirectoryServices.ActiveDirectory.ActiveDirectoryOperationException] { }
+
+    if (-not $domain) {
+        $result.changed = $true
```

Comment:
There's a lot of shared code here, it would be great if you could have the code below part of a function that each if statement calls.